### PR TITLE
Silent changes without value

### PR DESCRIPTION
**Pull request by rma4ok**
* This removes unnecessary tree update when switching from exact to rolling etc

### CODE REVIEW

**Comment by decrapifier** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Warnings</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:warning:</td>
      <td>The README has not been updated. Please update the README.</td>
    </tr>
  </tbody>
</table>


<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Messages</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:book:</td>
      <td>We were able to automatically fix some formatting issues in this PR for you!</td>
    </tr>
  
<tr>
      <td>:book:</td>
      <td>Could not find any browser results.</td>
    </tr>
  </tbody>
</table>


### Some things that were possibly fixed:

- Code that could be fixed via the --fix flag
- Formatting that could be fixed by prettier

Take a look at this commit to see what happened in detail: df9d7fbec26ecd0c75ae937b5b6263b7955216df


And look at this wiki page to see the reasoning behind the ESLint rules: https://github.com/smartprocure/eslint-config-smartprocure/wiki/Rules-and-Why-We-Chose-Them
<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 148cfaf550ccaf92b59d5169797312ad759c0f5c
</p>

**Review by daedalus28 — CHANGES_REQUESTED** [review]
We should try to make mutate smarter instead - if the new and prior values are hasValue false, suppress the changes. That approach is more general and might catch other cases we haven't thought of.

**Comment by decrapifier** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Warnings</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:warning:</td>
      <td>The README has not been updated. Please update the README.</td>
    </tr>
  </tbody>
</table>


<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Messages</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:book:</td>
      <td>Could not find any browser results.</td>
    </tr>
  </tbody>
</table>


<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 4fa3d008354e195976c75a074f773ee36aca57c3
</p>

File: src/node.js [hunk 1]
```diff
@@ -55,7 +55,10 @@ export let dedupeWalk = (fn, tree, { target = {}, dedupe } = {}) => {
   })(tree)
 }
 
-export let hasContext = node => node && node.context
+let isValue = x => !_.isNil(x) && (_.isNumber(x) || !_.isEmpty(x))
+
+export let hasContext = _.flow(_.get('context'), _.values, _.some(isValue))
```
Comment (daedalus28) [inline src/node.js:60]: This fundamentally changes the meaning of `hasContext` - the idea was to be a quick heuristic to determine if a node supports returning results, not whether it has any. Additionally, reactors shouldn't care about the values of context results

File: src/reactors.js [hunk 2]
```diff
@@ -45,16 +45,23 @@ export let reactors = {
   // ported from main app ¯\_(ツ)_/¯
   field: standardChange,
   type: standardChange,
-  mutate: (parent, node, event, reactor, types, lookup) =>
-    _.flow(
+  mutate(parent, node, event, reactor, types, lookup) {
+    if (
+      !hasValue(event.node) &&
+      !hasValue(event.previous) &&
+      !hasContext(event.node)
```
Comment (daedalus28) [inline src/reactors.js:52]: this check should be removed - nodes with contexts can and often do trigger other nodes to update (for example `facets`).

**Review by daedalus28 — CHANGES_REQUESTED** [review]

File: src/reactors.js (hunk 2 above)
Comment (rma4ok) [reply src/reactors.js:52]: Lots of tests are broken without it, what if a mutation happens inside context but hasValue is false? do we want to cancel that tree update?

**Pull request merged**